okhttp/src/main/java/okhttp3/internal/http/RealInterceptorChain.kt

```diff
@@ -120,11 +120,6 @@ class RealInterceptorChain(
           " must call proceed() exactly once")
     }
 
-    // Confirm that the intercepted response isn't null.
-    if (response == null) {
```

Comment:
I thought this was here to ensure nullability of what's returned from java callers.

Reply:
My understanding is that by this point in Kotlin, it can't be null.

Reply:
TIL - Kotlin null checks are a sham

Reply:
OK, so situation is

1) Without explicit null check in our code, the return value is enforced

```
    val response = interceptor.intercept(next)

    // Confirm that the next interceptor made its required call to chain.proceed().
    if (exchange != null && index + 1 < interceptors.size && next.calls != 1) {
      throw IllegalStateException("network interceptor " + interceptor +
          " must call proceed() exactly once")
    }

//    // Needed to enforce Kotlin null checks
//    @Suppress("SENSELESS_COMPARISON")
//    if (response == null) {
//      throw NullPointerException("interceptor $interceptor returned null")
//    }

//    if (response.body() == null) {
//      throw IllegalStateException(
//          "interceptor $interceptor returned a response with no body")
//    }

    return response
```

NPE

```
java.lang.NullPointerException
	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:124)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:115)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:90)
```

Reply:
With *later* explicit check

```
java.lang.NullPointerException: interceptor okhttp3.InterceptorTest$$Lambda$62/0x0000000800157c40@45752059 returned null
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:126)
	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:43)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:115)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:90)
```

Reply:
So Kotlin doesn't differentiate between a NPE without a message and with one.  But our test does.  So we are both right.

Reply:
Cool! 👍Good to learn regardless!